### DISCUSSION TITLE
Issues with compiling UMAT as .dll - Hard coding solution

### DISCUSSION THREAD

**Discussion by WaveHello** [Anura3D/Anura3D_OpenSource#181] [General]
I've had trouble compiling external soil models as .dll files and compiling them with the Anura3D solution. To get around compiling the .dll file it's possible to hard code the UMAT into the Anura3D source code.

 The purpose of this discussion is to detail and answer questions related to that process. The initiation of this conversation started in [another discussion thread](https://github.com/Anura3D/Anura3D_OpenSource/discussions/131#discussioncomment-8093205) and was moved here to make the discussion easier to find in the future.

Instructions to hardcode the UMAT in Anura3D can be found [here](https://github.com/WaveHello/Anura3D-Info/blob/main/Fortran_Info/Hard_Coding_ESM_Guide.md). 

Common problems:
* **Implicit typing**  - Some UMATs use implicit typing (eg.  ```IMPLICIT INTEGER (A-C)```- makes all variables that have a name starting with any letter A-C  an integer). Implicit typing was introduced early in Fortran history and has been deprecated in the Fortran standard. Implicit typing can make debugging challenging and obfuscate what certain code does. Therefore, the hard coding instructions above recommend removing implicit typing by typing ```implicit none``` at the top of the UMAT subroutines. This will cause some variables to become undefined. If the user wants to ignore dealing with implicit typing the easiest way is to not type implicit none and accept the debugging risk.

* **Collision of UMAT variable, subroutine, and function names with other subroutines already in the ```ExternalSoilModels.For``` file** - The best way to get around this is to follow the **Updated Hardcoding instructions** at the bottom of the guide. Creating a new module promotes data encapsulation and prevents names from conflicting.

**Note**: When I was first going through the process of  hardcoding a UMAT into Anura3D I didn't know how to create a new module to hold the UMAT and then use that module in the ```ExternalSoilModel.For``` file. Creating a module to hold the UMAT and using that module in the ```ExternalSoilModel.For``` file is highly recommended. Sometime in the future, I will update the hard coding instructions so that this is the standard procedure, for the time being some information is provided in the  **Updated Hardcoding instructions** section at the bottom of the guide.

For an creating a UMAT module and using it in the ```ExternalSoilModel.For``` file see:
1) [ExternalSoilModel.for file](https://github.com/WaveHello/Anura3D_Moore/blob/Anura3D_VelCorrection/src/MaterialModels/ExternalSoilModel.f90)
2) [Example UMAT Module](https://github.com/WaveHello/Anura3D_Moore/blob/Anura3D_VelCorrection/src/MaterialModels/HardcodedModels/ESM_MCSS.f90)

**Comment by WaveHello** [comment]
Hello @Marlon-li,

I  haven't implemented an cam-clay, however the procedure shouldn't depend on the constitutive model.  If you provide screen shots of the errors that you're getting I should be able to help you.

**Comment by Marlon-li** [comment]
[Barodesy_Abaqus_Plaxis.pdf](https://github.com/Anura3D/Anura3D_OpenSource/files/14004851/Barodesy_Abaqus_Plaxis.pdf)
[umat_BAROCL.f90.pdf](https://github.com/Anura3D/Anura3D_OpenSource/files/14004860/umat_BAROCL.f90.pdf)
Dear WaveHello,
I truly appreciate your selfless assistance. I apologize for the inconvenience of using images to indicate errors, so I have attached the umat code for the constitutive model that I found. Due to the inability to upload files with the .f90 extension, I changed the extension to .pdf. Simply remove the .pdf extension to view the files. I have also included the documentation for the code. Thank you very much for your help.
Best regards!

**Reply by WaveHello** [reply]
Hello Marlon,

If you're only using the umat subroutine (which is the one you shared with me) the problem is you also need the ESM subroutine. The ESM acts like a wrapper for the UMAT and is what interfaces with the rest of the source code. For an example of an ESM you can look at any of the soil models in Anura3D, including [Example UMAT Module](https://github.com/WaveHello/Anura3D_Moore/blob/Anura3D_VelCorrection/src/MaterialModels/HardcodedModels/ESM_MCSS.f90). If the Bardosie model came with an ESM the next steps are pretty straight forward. If it didn't, you can use one of the ESM subroutines in Anura3D as a template and modify it to your needs.

**Comment by linjiayu777** [comment]
Hello WaveHello,
Nice to meet you, i am working on the implementation of the sanisand-like constitutive model. I just following the instructions proveided by the offical doucumentation while it cant work well though it could be compiled successfully. Have you ever implemented the sanisand ? Any advice is appreciated and thank you very much!
Best regards 
jiayu

**Reply by WaveHello** [reply]
I haven't implemented sanisand. I'd recommend building out a flow chat of the constitutive model, to test the individual functions as you go along and use something like incremental driver to be able to test the constitutive model outside of Anura3D. Testing outside of Anura decreases the complexity and in my experience makes it faster to develop the initial model.

**Reply by linjiayu777** [reply]
thanks very much!

**Comment by 19Dario88** [comment]
Hi @WaveHello,
I'm trying to hardcode the MohrCoulomb_StrainSoft you developed into the ANURA3D (2024) code.
I followed this steps:
1. adding the f90 files (MohrCoulomb_StrainSoft.f90 and MCSS_funcs.f90) to the general code (fig.1); 
![fig1](https://github.com/user-attachments/assets/2fade81d-d91b-4472-a8bf-7a9120865769)
2. including the module into ModExternalSoilModel (fig.2);
![fig2](https://github.com/user-attachments/assets/d300fa34-94ae-4e1c-8e94-29452e8602bc)
3. including the module into ModReadMaterialData (fig.3);
![Fig3](https://github.com/user-attachments/assets/f4ce932b-e6f0-4c4f-b846-2ede3e10627a)
4. Defining the pointer to the ESM in ReadMaterialData.for (Fig.4);
![Fig4](https://github.com/user-attachments/assets/9d7b5213-daee-40fe-a93e-0bb3823521a2)
However, several error appear, and the code building stops (Fig.5).
![Fig5](https://github.com/user-attachments/assets/233aebaf-03d5-463c-80ca-4936379bdc11)
What am I wrong? Thank you in advance for any advice..
Dario

**Reply by WaveHello** [reply]
Hello,

I won't be able to provide much assistance with your question. I'm in the middle of a big project and don't have the availability. The only advice I can give is that you won't need the Reorder_Real_Array function. There might be a problem with the input real type to the reorder_real_array function. Reorder_Real_Array was made so that the voigt order that's used in Anura3D could be converted to the order using in incrementalDriver and as such isn't necessary when using the MCSS module in Anura3D.

**Reply by 19Dario88** [reply]
Hi, 
thank you for your precious advice, now it works fine!
Dario

**Reply by WaveHello** [reply]
Glad it worked! Good luck with everything!

**Reply by linjiayu777** [reply]
Hello,
I'm confused about your hardcode process, there may be some deviations between the instructions provided by WaveHello and yourself and your operation is more likely the official document(just my opinion). Do you modify the Globalconstants or MPMconvphase as he did? 
thanks in advance, and advice is well appreciated
jiayu

**Reply by 18052121369** [reply]
> Hi, thank you for your precious advice, now it works fine! Dario
@19Dario88 
Dear Dario,
I have some questions about how to hardcode the MohrCoulomb_StrainSoft you developed into the ANURA3D (2024) code. Did you successfully use the softening constitutive model to run your model ? I found that theESM_MCSS.f90 has been included in the hard-coded folder, but it is not work , I do not know the reason, or I do not know how to use it, can you share with me how you hard-coded Moore Coulomb softening model and how to use it ?
![Uploading 屏幕截图 2024-11-23 210418.png…]()

**Reply by AgustinToroF** [reply]
Hi @19Dario88. How did you solve it the issue about the "Reorder_Real_Array"?

**Reply by hafizullahge35** [reply]
> Hi, thank you for your precious advice, now it works fine! Dario

Hello Dario, 
Hope your're doing well. I am stuck with compiling the External Material Model (muI Rheology). I wrote the code in Fortran, converted it to a .dll, and placed it in the specified location, but it still isn't working. I am using Anura3D (2025). Help would be really appreciated.
Regards.

**Reply by panf7422-art** [reply]
> Hi,嗨， thank you for your precious advice, now it works fine!感谢您的宝贵建议，现在可以正常使用了！ Dario达里奥
Hello, I followed the settings shown in your screenshot above, and I also removed parts of the Reorder_Real_Array function in MohrCoulomb_StrainSoft.f90, but it still reports many errors. Could you please guide me on where the problem might be? My email is 2081231891@qq.com. If convenient, you can tell me your email, and I will send you screenshots of the specific content.

**Comment by 18052121369** [comment]
@WaveHello 
Dear WaveHello,
I am trying to use Anura to implement Mohr-Coulomb TrainSoftening.
1、First of all, I tried to generate the dll file according to the tutorial, but the error is as follows :
![image](https://github.com/user-attachments/assets/b54972ca-c307-46b2-882a-17933c3c100a)
2、Next, I found your suggestion in the discussion area, hard-coded MohrCoulomb _ StrainSoft.f90 into the Anura source code, and I tried several times according to Dario 's method here, and found that my results are the same as his, my error is very much.
①	adding the f90 files (MohrCoulomb_StrainSoft.f90 and MCSS_funcs.f90) to the general code (fig.2)
![image](https://github.com/user-attachments/assets/2f545fa7-4de6-4193-9e06-e7c67fa3a533)
②	including the module into ModExternalSoilModel (fig.3)
![image](https://github.com/user-attachments/assets/bb92bc57-5f66-4e4d-a863-49c65f62e643)
③	including the module into ModReadMaterialData (fig.4);
![image](https://github.com/user-attachments/assets/49d334dd-d393-4e2e-8bfd-72107d95d8d9)
④	Defining the pointer to the ESM in ReadMaterialData.for (Fig.5);
![image](https://github.com/user-attachments/assets/cced8b1f-bfd6-48d4-a7dc-02e2554b7197)
⑤	However, so many error appear, and the code building stops, I am very confused about this, I try to follow the method of Dario.
![image](https://github.com/user-attachments/assets/3474537d-df26-4e17-9585-ca892b4e4fd3)
3、Finally, I found that in your other forum Anura3D _ Moore-Anura3D _ VelCorrection, I released your code, and I found that there was a hardcode folder with ESM _ MCSS and ESM _ NAMC code in it. I don 't know whether this means that MCSS is hard-coded into Anura 's source code. When I used this code release to model Anura _ 2023.exe, I did not find any place to use this MCSS model in GID, or i just do not konw how to use it. 
Can you answer these questions ? I have been troubled by this problem for a long time and look forward to your reply.
![image](https://github.com/user-attachments/assets/6191e1ad-0cf7-4156-ae3e-e86df2eb2c93)

**Reply by AgustinToroF** [reply]
Hi. Did you solve it the issue to compile ESM on Anura3D?
